Cluster mode draft PR #2

barshaul opened the pull request:
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [ ] Does `$ tox` pass with this change (including linting)?
- [ ] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [ ] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change

_Please provide a description of the change here._

barshaul committed:
Added support for cluster mode

cr: https://code.amazon.com/reviews/CR-55243535

barshaul committed:
Added support for cluster mode

cr: https://code.amazon.com/reviews/CR-55243535

barshaul committed:
Merge branch 'andymccurdy:master' into cluster-mode

barshaul committed:
Changed crc to use implementation from python stdlib binascii.
Reformated files and sorted imports and static arrays.

barshaul committed:
Changed crc to use implementation from python stdlib binascii.
Reformated files and sorted imports and static arrays.

barshaul committed:
Merge from andymccurdy/redis-py

barshaul committed:
Merged branch 'andymccurdy-master' into cluster-mode

barshaul committed:
1. Added an optional argument of function pointer 'on_redis_connection' to override the Connection's class on_connect function.
2. Replaced CLusterConnection and SSLClusterConnection classes in RedisCluster.on_connect function that would be responsable of establishing readonly connection and setting the connection's parser to ClusterParser.
3. Added support in reading from replicas with Round-Robin load balancing.
4. Added readonly and readwrite commands.

barshaul committed:
1. Added an optional argument of function pointer 'on_redis_connection' to override the Connection's class on_connect function.
2. Replaced CLusterConnection and SSLClusterConnection classes in RedisCluster.on_connect function that would be responsable of establishing readonly connection and setting the connection's parser to ClusterParser.
3. Added support in reading from replicas with Round-Robin load balancing.
4. Added readonly and readwrite commands.

barshaul committed:
Added support for 'from_url' method

barshaul committed:
Added support for 'from_url' method

barshaul committed:
Fetch upstream and separate commands into smaller categories

barshaul committed:
Merge branch 'andymccurdy-master' into cluster-mode

barshaul committed:
Change f-string to str.format

barshaul committed:
Change f-string to str.format

barshaul committed:
fixings flake8 errors

barshaul committed:
fixings flake8 errors

barshaul committed:
1) Redis-py/master commands.py  merged.
 2) Commands class divided into smaller categories

barshaul committed:
1) Redis-py/master commands.py  merged.
 2) Commands class divided into smaller categories

barshaul committed:
1) Redis-py/master commands.py  merged.
 2) Commands class divided into smaller categories

barshaul committed:
1) Redis-py/master commands.py  merged.
 2) Commands class divided into smaller categories

barshaul committed:
Created a docker test environment for Redis Cluster with 2 primaries and 1 replica. Added Makefile target 'test-cluster' to run all the cluster's tests.

barshaul committed:
Created a docker test environment for Redis Cluster with 2 primaries and 1 replica. Added Makefile target 'test-cluster' to run all the cluster's tests.

chayim committed:
Pipeline DISCARD support (#1565)

closes #1539

Part of #1546

(cherry picked from commit febede19423c95515a7548cd73aa1a90c639ba1f)

cr: https://code.amazon.com/reviews/CR-57987360

chayim committed:
Pipeline DISCARD support (#1565)

closes #1539

Part of #1546

(cherry picked from commit febede19423c95515a7548cd73aa1a90c639ba1f)

cr: https://code.amazon.com/reviews/CR-57987360

barshaul committed:
1. Added tests for RedisCluster object.
2. Fixed bugs related to moved/cluster down error.
2. Added another replica for the cluster's docker env.

barshaul committed:
1. Added tests for RedisCluster object.
2. Fixed bugs related to moved/cluster down error.
2. Added another replica for the cluster's docker env.

barshaul committed:
Merged from upstream

barshaul committed:
Merge branch 'andymccurdy-master1' into cluster-mode

barshaul commented on docker-entry.sh:
```diff
@@ -15,5 +15,7 @@ if [ "${GITHUB_ACTIONS}" = true ] ; then
 # use the wait-for-it util to ensure the server is running before invoking Tox
-util/wait-for-it.sh ${REDIS_MASTER} -- tox -- --redis-url=redis://"${REDIS_MASTER}"/9
+util/wait-for-it.sh ${REDIS_MASTER} -- tox -- --redis-url=redis://"${REDIS_MASTER}"/${DB}
+#util/wait-for-it.sh ${REDIS_MASTER} -- tox -- tests/test_cluster.py::TestRedisClusterObj::test_moved_redirection_after_failover --redis-url=redis://"${REDIS_MASTER}"/${DB}
```
For debug purposes, will be deleted.

barshaul commented on tests/test_sentinel.py:
```diff
@@ -64,156 +84,129 @@ def timeout_if_down(self, node):
-def test_reset(cluster, sentinel):
-    cluster.master['is_odown'] = True
-    assert sentinel.sentinel_reset('mymaster')
+    def test_discover_master(sentinel, master_ip):
```
I  moved all tests to be under a class so I could use 'skip_if_cluster_mode' on all together.

barshaul committed:
Added support in pub-sub

barshaul committed:
Added support in pub-sub

barshaul committed:
Migrated the cluster tests into the existing target (so 'make test' will cover both strict and cluster redis) and suppressed the user warnings.

barshaul committed:
Migrated the cluster tests into the existing target (so 'make test' will cover both strict and cluster redis) and suppressed the user warnings.

barshaul committed:
Added ClusterPubSub tests

barshaul committed:
Added ClusterPubSub tests

barshaul committed:
Added tests for RedisCluster and NodesManager

barshaul committed:
Added tests for RedisCluster and NodesManager

barshaul committed:
Enabled existing tests for cluster mode: multiprocessing, connection pool. Added support in 'ping' command to be sent to all nodes.

barshaul committed:
Enabled existing tests for cluster mode: multiprocessing, connection pool. Added support in 'ping' command to be sent to all nodes.

chayim commented on docker-compose.yml:
```diff
@@ -34,13 +34,45 @@ services:
       - "26381:26381"
 
+  cluster_master_1:
+    build: docker/cluster/master_1
```
Let's use the container_name clause for each docker (even though we're not elsewhere). Locally, this makes debugging significantly easier.

barshaul replied:
Added

chayim commented on docker-compose.yml:
```diff
@@ -34,13 +34,45 @@ services:
+  cluster_master_1:
+    build: docker/cluster/master_1
+    ports:
+      - "7000:7000"
```
Since "regular" redis master/slave claimed 6379/6380, and sentinel is standardized on 26379/26380+, let's keep this in theme and use *either* 46379+ or 16379+.

barshaul replied:
Changed to 16379+

chayim commented on docker/cluster/master_1/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+COPY redis.conf /usr/local/etc/redis/redis.conf
+COPY nodes.conf /usr/local/etc/redis/nodes.conf
+
+EXPOSE 7000
```
You'll need to use the same port changes here.

barshaul replied:
done

chayim commented on docker/cluster/master_1/Dockerfile:
```diff
@@ -0,0 +1,9 @@
+
+EXPOSE 7000
+
+RUN chmod 666 /usr/local/etc/redis/nodes.conf
```
Why 666? I'd assume 0600, or 0660 at the least.

barshaul replied:
Actually, 006 does the job. I changed it.

chayim commented on redis/client.py:
```diff
@@ -1138,14 +1139,16 @@ class PubSub:
         # to lookup channel and pattern names for callback handlers.
-        self.encoder = self.connection_pool.get_encoder()
+        self.encoder = encoder
+        if self.encoder is None:
```
PERFECT

chayim commented on redis/utils.py:
```diff
@@ -36,3 +36,41 @@ def str_if_bytes(value):
+    """
+    merged = {}
+
+    for d in dicts:
```
I'm torn. I really don't think you should have to sanitise inputs to a class that merges dicts, and should be able to remove the _if not isinstance_ check. Is there a strong reason you feel the need?

barshaul replied:
removed

chayim commented on redis/utils.py:
```diff
@@ -36,3 +36,41 @@ def str_if_bytes(value):
+    This command is used when sending a command to multiple nodes
+    and they result from each node should be merged into a single list.
+    """
+    if not isinstance(res, dict):
```
Same idea here. I'm not sure you should have to check the object. I'd fully support a docstring change to be specific even type hints. Any reason for this preference?

barshaul replied:
removed

chayim commented on redis/utils.py:
```diff
@@ -36,3 +36,41 @@ def str_if_bytes(value):
+    and they result from each node should be merged into a single list.
+    """
+    if not isinstance(res, dict):
+        raise ValueError('Value should be of dict type')
```
```suggestion
        raise ValueError('Value should be a dict not {}'.format(str(type(res))))
```

chayim commented on redis/utils.py:
```diff
@@ -36,3 +36,41 @@ def str_if_bytes(value):
+    if not isinstance(res, dict):
+        raise ValueError('Value should be of dict type')
+
+    result = set([])
```
```suggestion
    result = set([])
```

chayim commented on redis/utils.py:
```diff
@@ -36,3 +36,41 @@ def str_if_bytes(value):
+
+    result = set([])
+
+    for _, v in res.items():
```
```suggestion
    for v in res.values():
        for value in v:
            result.add(value)
```

barshaul replied:
changed

chayim commented on redis/exceptions.py:
```diff
@@ -84,3 +84,81 @@ class AuthenticationWrongNumberOfArgsError(ResponseError):
+
+
+class RedisClusterException(Exception):
+    """
```
Nit: docstrings please

barshaul replied:
removed

chayim commented on redis/connection.py:
```diff
@@ -577,7 +605,12 @@ def connect(self):
         self._sock = sock
         try:
-            self.on_connect()
+            if self.redis_connect_func is not None:
```
Since the majority/default case is that the redis_connect_func is None, we should prioritize the default connection, and instead call the func only if defined.

barshaul replied:
Changed.

chayim commented on redis/connection.py:
```diff
@@ -503,7 +526,7 @@ def __init__(self, host='localhost', port=6379, db=0, password=None,
                  parser_class=DefaultParser, socket_read_size=65536,
                  health_check_interval=0, client_name=None, username=None,
-                 retry=None):
+                 retry=None, redis_connect_func=None):
```
Note to self: this class is getting long in the tooth.

chayim commented on redis/connection.py:
```diff
@@ -493,6 +502,20 @@ def read_response(self):
     DefaultParser = PythonParser
 
 
+class ClusterParser(DefaultParser):
```
IMHO this really belongs in redis/cluster.py. Since it's only used there, it keeps the cluster code nicely self-contained.

barshaul replied:
agreed, moved to cluster.py.

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+    TimeoutError,
+    TryAgainError,
+)
+from redis.utils import dict_merge, list_keys_to_dict, merge_result, \
```
Let's match the style already present:
from redis.utils import (
    dict_merge,
    list_keys_to_dict
    ....
)

barshaul replied:
done

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+    Result callback, handles different return types
+    switchable by the `aggregate` flag.
+    """
+    aggregate = options.get('aggregate', True)
```
if options.get('aggregate', True) is False:
    return res
return merge_result(command, res)

barshaul replied:
it's completely removed on the next commits

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+    Result callback, handles different return types
+    switchable by the `aggregate` flag.
+    """
+    aggregate = options.get('aggregate', True)
```
```suggestion
    if options.get('aggregate', True) is not True:
        return res
```

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+    Result callback, handles different return types
+    switchable by the `aggregate` flag.
+    """
+    aggregate = options.get('aggregate', True)
```
Same as sabove

barshaul replied:
removed on the next commits

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+                numsub_d[channel] = numsubbed
+
+    ret_numsub = []
+    for channel, numsub in numsub_d.items():
```
This should be a list comprehension

barshaul replied:
changed

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+
+def parse_cluster_nodes(resp, **options):
+    """
+    @see: http://redis.io/commands/cluster-nodes  # string
```
Thank you :+1:

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+RANDOM = "random"
+SLOT_ID = "slot-id"
+
+REDIS_ALLOWED_KEYS = (
```
Really good idea!

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+
+# Not complete, but covers the major ones
+# https://redis.io/commands
+READ_COMMANDS = frozenset([
```
I'll happily update this list for you, if you can somehow give me write access.

barshaul replied:
I sent you a collaborator access invitation

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+
+        # Get the startup node/s
+        from_url = False
+        if not startup_nodes:
```
PEP suggests we should actively None check like below. However, given your use case, maybe startup_nodes should default to an empty list. WDYT?

barshaul replied:
changed

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+                    "A ``db`` querystring option can only be 0 in cluster mode"
+                )
+            kwargs.update(url_options)
+            startup_nodes.append(ClusterNode(kwargs['host'], kwargs['port']))
```
IMHO this doesn't cover all cases. Note that host can be None, port defaults to a value, and from_url defaults to None. I think there should be an else with a raise.

barshaul replied:
it was covered by this check: ` if not startup_nodes:`
but I changed it to` 'else: raise error'`

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+                )
+            kwargs.update(url_options)
+            startup_nodes.append(ClusterNode(kwargs['host'], kwargs['port']))
+        if not startup_nodes:
```
I prefer the explicit syntax:
```suggestion
        if startup_nodes == []:
```

Alternatively a check on len(startup_nodes) == 0 would be nice. It helps juniors understand the code base, while achieving the same value.

barshaul replied:
changed to len(startup_nodes) == 0

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
```
Nice to see thought out ContextManagers

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+    def get_node(self, node_name=None, host=None, port=None):
+        return self.nodes_manager.get_node(node_name, host, port)
+
+    def get_all_primaries(self):
```
s/get_all_primaryies/get_primaries

I don't think we're going to want only some of the primaries. If someone does.. they can me a slice.

barshaul replied:
changed

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+    def get_all_primaries(self):
+        return self.nodes_manager.get_nodes_by_server_type(PRIMARY)
+
+    def get_all_replicas(self):
```
Sam as get_primaries

barshaul replied:
changed

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+    def get_random_node(self):
+        return random.choice(list(self.nodes_manager.nodes_cache.values()))
+
+    def get_all_nodes(self):
```
Same. get_nodes

barshaul replied:
changed

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+        node_flag = self.nodes_flags.get(command)
+
+        if node_flag == RANDOM:
+            return [self.get_random_node()]
```
Changes to call the above, renamed functions
...

barshaul replied:
changed

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+        If it reaches the number of times, the command will raises
+        ClusterDownException.
+        """
+        if not args:
```
IMHO this case really should never occur. We don't check in redis-py elsewhere and assume args[0]

barshaul replied:
removed

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+        ClusterDownException.
+        """
+        if not args:
+            raise RedisClusterException("Unable to determine command to use")
```
This should probably be a ValueError since there's no command.

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+        self.initialize()
+
+    def get_node(self, node_name=None, host=None, port=None):
+        if node_name is None:
```
```suggestion
        if node_name is None and host is None and port is None:
```

barshaul replied:
Changed to:
```suggestion
        if node_name is None and (host is None or port is None):
```

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+                    "2. host and port"
+                )
+                return None
+            if host == "localhost":
```
elif host is not None and node_name is None

barshaul replied:
changed

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+        """
+        Gets a node that servers this hash slot
+        """
+        try:
```
Can we shorten the try/catch or modify is so that the minimal code that triggers the KeyError does the raise?  Alternatively a series of has_key checks could do a raise SlotNotCoveredError.

barshaul replied:
Changed:
```
    def get_node_from_slot(self, slot, read_from_replicas=False,
                           server_type=None):
        """
        Gets a node that servers this hash slot
        """
        if self._moved_exception:
            self._update_moved_slots()
        if read_from_replicas:
            # get the server index in a Round-Robin manner
            node_idx = self.get_read_load_balancer().get_server_index(
                slot, len(self.slots_cache[slot]))
        elif (
                server_type is None
                or server_type == PRIMARY
                or len(self.slots_cache[slot]) == 1
        ):
            # return a primary
            node_idx = 0
        else:
            # return a replica
            # randomly choose one of the replicas
            node_idx = random.randint(
                1, len(self.slots_cache[slot]) - 1)
        try:
            target_node = self.slots_cache[slot][node_idx]
        except KeyError:
            raise SlotNotCoveredError(
                'Slot "{0}" not covered by the cluster. '
                '"skip_full_coverage_check={1}"'.format(
                    slot, self._skip_full_coverage_check)
            )
        return target_node
```

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+        return target_node
+
+    def get_nodes_by_server_type(self, server_type):
+        return [
```
I love this use of a list comprehension!

chayim commented on redis/cluster.py:
```diff
@@ -0,0 +1,1249 @@
+        ]
+
+    def get_read_load_balancer(self):
+        if not self.read_load_balancer:
```
```suggestion
    if self.read_load_balancer is None:
```

barshaul replied:
Cahnged

chayim reviewed (changes requested):
Amazing work @barshaul. I added some comments throughout for you.

barshaul committed:
Changes in the docker envs: Added contatiner name to each docker, changes the cluster's ports to 16379+, changed chmod for nodes.conf to 006

barshaul committed:
Changes in the docker envs: Added contatiner name to each docker, changes the cluster's ports to 16379+, changed chmod for nodes.conf to 006

barshaul committed:
Code refactoring and cleaning

barshaul committed:
Code refactoring and cleaning

barshaul committed:
Merge branch 'andymccurdy:master' into cluster-mode

barshaul committed:
Fixed 'COMMAND' response value to a dictionary

barshaul committed:
Fixed 'COMMAND' response value to a dictionary

barshaul committed:
Added CommandsParser to cache all commands info and retrieve keys from requested commands

cr: https://code.amazon.com/reviews/CR-57976849

barshaul committed:
Added CommandsParser to cache all commands info and retrieve keys from requested commands

cr: https://code.amazon.com/reviews/CR-57976849

barshaul committed:
Merge branch 'andymccurdy:master' into cluster-mode

barshaul committed:
Fixed CommandsParser relevant bugs

barshaul committed:
Fixed CommandsParser relevant bugs

cr: https://code.amazon.com/reviews/CR-57986762

barshaul committed:
Fixed CommandsParser relevant bugs

barshaul committed:
Changed the CommandsParser logic: if command is flagged as moveable, we will execute 'COMMAND GETKEYS' command to get from the server the key positions. if is a pubsub command, we'll use the hardcoded positions. else, we will use the start/last key and step count of 'COMMAND's output.

barshaul committed:
Changed the CommandsParser logic: if command is flagged as moveable, we will execute 'COMMAND GETKEYS' command to get from the server the key positions. if is a pubsub command, we'll use the hardcoded positions. else, we will use the start/last key and step count of 'COMMAND's output.

barshaul committed:
Changed the CommandsParser logic: if command is flagged as moveable, we will execute 'COMMAND GETKEYS' command to get from the server the key positions. if is a pubsub command, we'll use the hardcoded positions. else, we will use the start/last key and step count of 'COMMAND's output.

barshaul committed:
Merge branch 'andymccurdy:master' into cluster-mode

barshaul committed:
Added locks for thread safety

barshaul committed:
Added locks for thread safety

barshaul committed:
Removed skip_full_coverage_check and added require_full_coverage which set to True by default. With this changes, RedisCluster creation doesn't need to run cluster_require_full_coverage() by default, which didn't work with clusters without the CONFIG command, and will be called only if require_full_coverage was set to False by the user and the slots are not fully covered.

barshaul committed:
Removed skip_full_coverage_check and added require_full_coverage which set to True by default. With this changes, RedisCluster creation doesn't need to run cluster_require_full_coverage() by default, which didn't work with clusters without the CONFIG command, and will be called only if require_full_coverage was set to False by the user and the slots are not fully covered.

barshaul committed:
Added partial documentation for Cluster Mode support in the README file

barshaul committed:
Added partial documentation for Cluster Mode support in the README file

barshaul committed:
Fixed a small bug

barshaul committed:
Added support for passing 'redis_connect_func' argument in RedisCluster which will be called by RedisCluster function 'on_connect' when connection is getting established

barshaul committed:
Added support for passing 'redis_connect_func' argument in RedisCluster which will be called by RedisCluster function 'on_connect' when connection is getting established

barshaul committed:
1. Files/variables with 'slave' were renamed to 'replica'
2. Fixed the cluster's response_callbacks to be called only with cluster commands

barshaul committed:
1. Files/variables with 'slave' were renamed to 'replica'
2. Fixed the cluster's response_callbacks to be called only with cluster commands

barshaul committed:
Code cleanup

barshaul committed:
Code cleanup

asheryerm committed:
Add support for multi-key and cluster management commands in cluster-mode

asheryerm committed:
Add support for multi-key and cluster management commands in cluster-mode

asheryerm committed:
Add support for pipeline in cluster-mode

asheryerm committed:
Add support for pipeline in cluster-mode

asheryerm committed:
Merge pull request #5 from asheryerm/cluster-mode

Cluster mode

barshaul committed:
Added & fixed tests for Cluster Commands

barshaul committed:
Added & fixed tests for Cluster Commands

barshaul committed:
Merging upstream

barshaul committed:
Merge branch 'redis-master' into cluster-mode

barshaul committed:
Added RedisCluster documentation to the README file

barshaul committed:
Added RedisCluster documentation to the README file

asheryerm committed:
Support additional cluster management commands in cluster-mode

asheryerm committed:
Support additional cluster management commands in cluster-mode

asheryerm committed:
Fix mget_nonatomic

asheryerm committed:
Fix mget_nonatomic

asheryerm committed:
Merge pull request #8 from asheryerm/cluster-mode

Cluster mode

barshaul committed:
Added support in passing node flags as target nodes: PRIMARIES, REPLICAS, ALL_NODES and RANDOM. It  intends to replace executing commands with specifying directly target nodes (for example r.ping(r.get_primaries())), so the nodes will be internally resolved and the client will be able to retry after a topology change.

barshaul committed:
Added support in passing node flags as target nodes: PRIMARIES, REPLICAS, ALL_NODES and RANDOM. It  intends to replace executing commands with specifying directly target nodes (for example r.ping(r.get_primaries())), so the nodes will be internally resolved and the client will be able to retry after a topology change.

barshaul committed:
Merge branch 'cluster-mode' of https://github.com/barshaul/redis-py into cluster-mode

barshaul committed:
Added skip_full_coverage_check parameter to support skipping configuration check for 'cluster-require-full-coverage' if the user requested that the cluster won't require full coverage.

barshaul committed:
Added skip_full_coverage_check parameter to support skipping configuration check for 'cluster-require-full-coverage' if the user requested that the cluster won't require full coverage.

barshaul committed:
Changed the LoadBalancer to hold a primary_name to server_idx dict instead of slot to server_idx.
Added debug mode to print the executed command and target node.

barshaul committed:
Changed the LoadBalancer to hold a primary_name to server_idx dict instead of slot to server_idx.
Added debug mode to print the executed command and target node.

barshaul committed:
small fix

barshaul committed:
small fix

barshaul committed:
Changed ports

barshaul committed:
Merging master branch

barshaul committed:
Merging master branch

barshaul committed:
Merge branch 'master' of https://github.com/redis/redis-py into cluster-mode

barshaul committed:
Pull branch cluster mode from barshaul-redis-py

barshaul committed:
Removed all single dockers for Redis Cluster nodes and created a single docker that will create all nodes locally.

barshaul committed:
Small fix

barshaul committed:
Merge branch 'redis:master' into cluster-mode

barshaul committed:
Added attribution for redis-py-cluster

barshaul committed:
Small flake8 fix

barshaul committed:
Skip pattern PUBSUB test for cluster mode
